torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -0,0 +1,390 @@
+#include "ProcessGroupGloo.hpp"
+
+#include <gloo/allreduce_halving_doubling.h>
+#include <gloo/rendezvous/context.h>
+
+#define GENERATE_ALL_TYPES(type, func, args...)                         \
+  switch (type) {                                                       \
+    case ::at::ScalarType::Float: func<float>(args); break;             \
+    case ::at::ScalarType::Double: func<double>(args); break;           \
+    case ::at::ScalarType::Half: func<gloo::float16>(args); break;      \
+    case ::at::ScalarType::Char: func<int8_t>(args); break;             \
+    case ::at::ScalarType::Byte: func<uint8_t>(args); break;            \
+    case ::at::ScalarType::Int: func<int32_t>(args); break;             \
+    case ::at::ScalarType::Long: func<int64_t>(args); break;            \
+    default:                                                            \
+      throw std::runtime_error("Invalid scalar type");                  \
+  }
+
+namespace c10d {
+
+using KeyType = AlgorithmKey;
+using EntryType = std::unique_ptr<AlgorithmEntry>;
+
+namespace {
+
+// Wrap c10d store as Gloo store
+class GlooStore : public ::gloo::rendezvous::Store {
+ public:
+  GlooStore(const std::shared_ptr<::c10d::Store>& store)
+      : store_(store) {
+  }
+
+  void set(const std::string& key, const std::vector<char>& value) override {
+    // @pietern was unable to figure out a way to cast the value of
+    // type std::vector<char> to an std::vector<unsigned char>,
+    // so going with a copy here (it's not a hot path so it's fine).
+    std::vector<uint8_t> tmp(value.size());
+    memcpy(tmp.data(), value.data(), value.size());
+    store_->set(key, tmp);
+  }
+
+  std::vector<char> get(const std::string& key) override {
+    // @pietern was unable to figure out a way to cast the value of
+    // type std::vector<unsigned char> to an std::vector<char>,
+    // so going with a copy here (it's not a hot path so it's fine).
+    auto value = store_->get(key);
+    std::vector<char> tmp(value.size());
+    memcpy(tmp.data(), value.data(), value.size());
+    return tmp;
+  }
+
+  void wait(const std::vector<std::string>& keys) override {
+    store_->wait(keys, Store::kDefaultTimeout);
```

Comment:
I found the defaultTimeout is zero(), but in Util.hpp, I had to make it -1 to be compatible with socket timeout options etc.  Shall we keep it to be the same, wdyt?

Reply:
What do you mean, do you plug this directly into setsocktop?

Reply:
I plug kNoTimeout.count() directly into socket option, and it returns -1 since kNoTimeout was by default constructed with -1.